File: core/src/main/java/io/github/shaksternano/borgar/core/util/Environment.java
package io.github.shaksternano.borgar.core.util;

import io.github.shaksternano.borgar.core.Main;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Environment {

    private static final Map<String, String> customEnvVars = new HashMap<>();

    public static void load(File file) throws IOException {
        for (var line : Files.readAllLines(file.toPath())) {
            if (!line.isBlank()) {
                var envVar = line.split("=", 2);
                if (envVar.length == 2) {
                    setEnvVar(envVar[0].trim(), envVar[1].trim());
                } else {
                    Main.getLogger().error("Invalid environment variable: " + line);
                }
            }
        }
    }

    public static Optional<String> getEnvVar(String key) {
        var customEnvVar = customEnvVars.get(key);
        if (customEnvVar == null) {
            return Optional.ofNullable(System.getenv(key));
        } else {
            return Optional.of(customEnvVar);
        }
    }

    public static void setEnvVar(String key, String value) {
        customEnvVars.put(key, value);
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/util/Environment.kt
package io.github.shaksternano.borgar.core.util

import io.github.shaksternano.borgar.core.io.IO_DISPATCHER
import io.github.shaksternano.borgar.core.logger
import kotlinx.coroutines.withContext
import java.nio.file.Path
import kotlin.io.path.forEachLine
import kotlin.system.exitProcess

private val customEnvVars: MutableMap<String, String> = HashMap()

suspend fun loadEnv(path: Path) {
    try {
        withContext(IO_DISPATCHER) {
            path.forEachLine {
                if (it.isNotBlank()) {
                    val envVar = it.split("=", limit = 2)
                    if (envVar.size == 2) {
                        val key = envVar[0].trim()
                        val value = envVar[1].trim()
                        if (key.isNotBlank() && value.isNotBlank()) {
                            setEnvVar(key, value)
                        }
                    } else {
                        logger.error("Invalid environment variable: $it")
                    }
                }
            }
        }
    } catch (t: Throwable) {
        logger.error("Error reading environment variables from ${path.toAbsolutePath()}", t)
        exitProcess(1)
    }
}

fun getEnvVar(key: String): String? =
    customEnvVars[key] ?: System.getenv(key)

fun setEnvVar(key: String, value: String) {
    customEnvVars[key] = value
}
